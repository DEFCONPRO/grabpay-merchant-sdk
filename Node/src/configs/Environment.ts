import { TCountry, TEnvironment } from '../types';

export type EnvironmentSet = {
  baseUrl: string;
};

const definedEnvs: Partial<Record<TCountry, Record<TEnvironment, EnvironmentSet>>> = {
  VN: {
    STG: {
      baseUrl: 'https://stg-paysi.moca.vn',
    },
    PRD: {
      baseUrl: 'https://partner-gw.moca.vn',
    },
  },
};

const fallbackEnvironment: Record<TEnvironment, EnvironmentSet> = {
  STG: {
    baseUrl: 'https://partner-api.stg-myteksi.com',
  },
  PRD: {
    baseUrl: 'https://partner-api.grab.com',
  },
};

export function getEnvs(country: TCountry, env: TEnvironment) {
  const countryEnv = country in definedEnvs ? definedEnvs[country] : undefined;
  return countryEnv ? countryEnv[env] : fallbackEnvironment[env];
}
