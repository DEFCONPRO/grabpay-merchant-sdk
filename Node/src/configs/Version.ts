export const SDK_VERSION = '2.0.0';
export const SDK_SIGNATURE = '6EB66646DBF103DC114E34AE6E01C261A217A820357C3B08F3D4E7D4475853C7';
