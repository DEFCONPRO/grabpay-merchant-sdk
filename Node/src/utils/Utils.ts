import { THttpMethod } from '../types';
const Buffer = require('buffer/').Buffer;
import { createHash, createHmac, randomBytes } from 'crypto';

const btoa = (str: string) => Buffer.from(str).toString('base64');

export function base64URLEncode(str: string) {
  return str.replace(/=/g, '').replace(/\+/g, '-').replace(/\//g, '_');
}

export type GenerateHmacParams = {
  httpMethod: THttpMethod;
  requestUrl: string;
  contentType: string;
  timestamp: string;
  requestBody: string;
  secret: string;
};

export function hashSha256ToBase64(message: string) {
  return createHash('sha256').update(message, 'utf-8').digest('base64');
}

export function hmacSha256ToBase64(message: string, secret: string) {
  return createHmac('sha256', secret).update(message, 'utf-8').digest('base64');
}

export function generateHmac({
  httpMethod,
  requestUrl,
  contentType,
  timestamp,
  requestBody,
  secret,
}: GenerateHmacParams) {
  let hashedPayload = '';
  if (httpMethod !== 'GET' && requestBody) {
    hashedPayload = hashSha256ToBase64(requestBody);
  }

  const requestData = [httpMethod, contentType, timestamp, requestUrl, hashedPayload, ''].join('\n');

  return hmacSha256ToBase64(requestData, secret);
}

export function generatePOPSignature(accessToken: string, clientSecret: string) {
  const timestamp = new Date();
  const timestampUnix = Math.round(timestamp.getTime() / 1000);
  const message = timestampUnix.toString() + accessToken;
  const signature = hmacSha256ToBase64(message, clientSecret);
  const sub = base64URLEncode(signature);
  const payload = {
    sig: sub,
    time_since_epoch: timestampUnix,
  };
  const payloadBytes = JSON.stringify(payload);

  return base64URLEncode(btoa(payloadBytes));
}

const DEFAULT_RANDOM_STRING_CHAR_SET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';

export function generateRandomString(length: number, charSet: string = DEFAULT_RANDOM_STRING_CHAR_SET) {
  const buf = randomBytes(length);
  return [...buf.values()]
    .map((bufValue) => Math.floor((bufValue / 256) * charSet.length))
    .map((charIdx) => charSet.charAt(charIdx))
    .join('');
}

export function generateMsgId() {
  return 'xxxxxxxxxxxx4xxxyxxxxxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
    const r = (Math.random() * 16) | 0,
      v = c == 'x' ? r : (r & 0x3) | 0x8;
    return v.toString(16);
  });
}

export function replaceUrl(url: string, args: Record<string, string>) {
  const regex = /{([a-zA-Z]+)}/gm;

  return url.replace(regex, (matches, item) => args[item] || '');
}

export function getGMTString() {
  return new Date().toUTCString();
}
