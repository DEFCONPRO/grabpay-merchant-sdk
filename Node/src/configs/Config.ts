import { TCountry, TEnvironment } from '../types';
import { ApiUrlSet, getUrls } from './ApiUrls';
import { EnvironmentSet, getEnvs } from './Environment';

const definedCountryScopes: Partial<Record<TCountry, string>> = {
  VN: 'payment.vn.one_time_charge',
};
const fallbackScope = ['openid', 'payment.one_time_charge'].join(' ');

type CommonClientRawConfig = {
  env: TEnvironment;
  country: TCountry;
  partnerId: string;
  partnerSecret: string;
  merchantId: string;
};

export type OfflineClientRawConfig = CommonClientRawConfig & {
  terminalId: string;
};

export type OnlineClientRawConfig = CommonClientRawConfig & {
  clientId: string;
  clientSecret: string;
  redirectUrl: string;
};

export type ClientRawConfig = OnlineClientRawConfig | OfflineClientRawConfig;

export type InitializedConfig<T extends ClientRawConfig> = {
  envVars: EnvironmentSet;
  urlSet: ApiUrlSet;
  scope: string;
} & T;

export type OnlineClientConfig = InitializedConfig<OnlineClientRawConfig>;
export type OfflineClientConfig = InitializedConfig<OfflineClientRawConfig>;
export type ClientConfig = OnlineClientConfig | OfflineClientConfig;

export function initializeWithConfig<T extends ClientRawConfig>(input: T): InitializedConfig<T> {
  return {
    ...input,
    envVars: getEnvs(input.country, input.env),
    urlSet: getUrls(input.country),
    scope: (input.country in definedCountryScopes ? definedCountryScopes[input.country] : undefined) || fallbackScope,
  };
}
