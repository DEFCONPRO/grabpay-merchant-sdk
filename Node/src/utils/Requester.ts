import axios, { AxiosInstance, AxiosRequestConfig } from 'axios';
import { generateHmac, getGMTString } from './Utils';
import { SDK_SIGNATURE, SDK_VERSION } from '../configs/Version';
import { ClientConfig } from '../configs/Config';
import { THttpMethod } from '../types';

export interface IRequester {
  hmacRequest<TResponseBody, TRequestBody>(
    request: Readonly<RequestParams<TRequestBody>>,
  ): Promise<Response<TResponseBody>>;
  request<TResponseBody, TRequestBody>(
    request: Readonly<RequestParams<TRequestBody>>,
  ): Promise<Response<TResponseBody>>;
}

export type RequestParams<TRequestBody> = {
  path: string;
  method: THttpMethod;
  query?: Record<string, string>;
  headers?: Record<string, string>;
} & ({ body: string } | { data: TRequestBody } | {});

export type Response<TResponseBody> = {
  status: number;
  headers: Record<string, string>;
  data: TResponseBody;
};

function buildHeadersForHmacRequest<TRequestBody>(
  requestParams: Readonly<RequestParams<TRequestBody>>,
  partnerId: string,
  partnerSecret: string,
): Record<string, string> {
  const { headers = {}, path, method, query } = requestParams;
  const contentType = headers['Content-Type'] || 'application/json';
  const timestamp = headers['Date'] || getGMTString();
  let pathWithParams = path;
  if (query && Object.keys(query).length > 0) {
    pathWithParams = `${pathWithParams}?${new URLSearchParams(query).toString()}`;
  }
  const hmac = generateHmac({
    contentType,
    httpMethod: method || 'GET',
    requestBody: 'body' in requestParams ? requestParams.body : '',
    requestUrl: pathWithParams,
    secret: partnerSecret,
    timestamp,
  });
  return {
    Authorization: `${partnerId}:${hmac}`,
    'Content-Type': contentType,
    Date: timestamp,
    ...headers,
  };
}

export default class AxiosRequester implements IRequester {
  _config: ClientConfig;
  _requester: AxiosInstance;

  constructor(config: ClientConfig) {
    const { baseUrl } = config.envVars;
    this._requester = axios.create({ baseURL: baseUrl, transformRequest: [] });
    this._config = config;
  }

  hmacRequest<TResponseBody, TRequestBody>(requestParams: Readonly<RequestParams<TRequestBody>>) {
    const { path, method, query } = requestParams;
    let body = '';
    if ('body' in requestParams) {
      body = requestParams.body;
    } else if ('data' in requestParams) {
      body = JSON.stringify(requestParams.data);
    }
    const hmacSignedHeaders = buildHeadersForHmacRequest(
      {
        ...requestParams,
        body,
      },
      this._config.partnerId,
      this._config.partnerSecret,
    );

    return this.request<TResponseBody, never>({
      headers: hmacSignedHeaders,
      body,
      method,
      path,
      query,
    });
  }

  async request<TResponseBody, TRequestBody>(
    requestParams: Readonly<RequestParams<TRequestBody>>,
  ): Promise<Response<TResponseBody>> {
    let body = '';
    if ('body' in requestParams) {
      body = requestParams.body;
    } else if ('data' in requestParams) {
      body = JSON.stringify(requestParams.data);
    }
    const axiosConfig: AxiosRequestConfig = {
      method: requestParams.method,
      url: requestParams.path,
      headers: {
        ...requestParams.headers,
        'X-Sdk-Country': this._config.country,
        'X-Sdk-Language': 'nodejs',
        'X-Sdk-Signature': SDK_SIGNATURE,
        'X-Sdk-Version': SDK_VERSION,
      },
      ...(requestParams.method === 'GET' ? {} : { data: body }),
      ...(requestParams.query && Object.keys(requestParams.query).length > 0 ? { params: requestParams.query } : {}),
    };

    try {
      const response = await this._requester.request<TResponseBody>(axiosConfig);
      return {
        status: response.status,
        headers: response.headers,
        data: response.data,
      };
    } catch (err) {
      if (axios.isAxiosError(err) && err.response) {
        return {
          status: err.response.status,
          headers: err.response.headers,
          data: err.response.data,
        };
      }
      throw err;
    }
  }
}
