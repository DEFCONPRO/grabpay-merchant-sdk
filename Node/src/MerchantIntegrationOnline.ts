import Requester from './utils/Requester';
import { SDK_VERSION } from './configs/Version';
import { initializeWithConfig, OnlineClientConfig, OnlineClientRawConfig } from './configs/Config';
import { TCurrency, TPaymentMethod, TTxStatus } from './types';
import * as Utils from './utils/Utils';

export type ChargeInitParams = {
  partnerTxID: string;
  partnerGroupTxID: string;
  amount: number;
  currency: TCurrency;
  description: string;
  hidePaymentMethods?: TPaymentMethod[];
  metaInfo?: {
    brandName?: string;
    location?: Partial<{
      ipAddress: string;
      longitude: number;
      latitude: number;
      accuracy: number;
    }>;
    device?: Partial<{
      deviceID: string;
      deviceModel: string;
      deviceBrand: string;
      iosUDID: string;
      imei: string;
    }>;
    subMerchant?: Partial<{
      name: string;
      category: string;
    }>;
    partnerUserInfo?: Partial<{
      id: string;
      phone: string;
      email: string;
    }>;
    echo?: string;
  };
  items?: {
    supplierName?: string;
    itemName?: string;
    category?: string;
    quantity?: number;
    price?: number;
    itemCategory?: string;
    url?: string;
    imageURL?: string;
    supplierDetails?: {
      supplierID?: string;
      supplierName?: string;
      supplierURL?: string;
      legalID?: string;
      supplierPhone?: string;
      supplierEmail?: string;
      supplierAddress?: string;
      supplierRating?: number;
    };
    shippingDetails?: {
      firstName?: string;
      lastName?: string;
      address?: string;
      city?: string;
      postalCode?: string;
      phone?: string;
      email?: string;
      countryCode?: string;
    };
  }[];
};

export type ChargeInitResponse = {
  partnerTxID: string;
  request: string;
};

export type GetOAuth2TokenParams = {
  code: string;
  codeVerifier: string;
};

export type GetOAuth2TokenResponse = {
  access_token: string;
  token_type: string;
  expires_in: number;
  id_token: string;
};

export type ChargeCompleteParams = {
  partnerTxID: string;
  accessToken: string;
};

export type ChargeCompleteResponse = {
  partnerTxID: string;
};

export type GetChargeStatusParams = {
  partnerTxID: string;
  currency: TCurrency;
  accessToken: string;
};

export type GetChargeStatusResponse = {
  txID: string;
  paymentMethod: TPaymentMethod;
  /**
   * @deprecated
   */
  status: string;
  description: string;
  txStatus: TTxStatus;
  reason: string;
};

export type RefundParams = {
  partnerTxID: string;
  partnerGroupTxID: string;
  amount: number;
  currency: TCurrency;
  originTxID: string;
  description: string;
  accessToken: string;
};

export type RefundResponse = {
  txID: string;
  /**
   * @deprecated
   */
  status: string;
  paymentMethod: TPaymentMethod;
  description: string;
  txStatus: TTxStatus;
  reason: string;
};

export type GetRefundStatusParams = {
  partnerTxID: string;
  currency: TCurrency;
  accessToken: string;
};

export type GetRefundStatusResponse = {
  txID: string;
  /**
   * @deprecated
   */
  status: string;
  paymentMethod: TPaymentMethod;
  description: string;
  txStatus: TTxStatus;
  reason: string;
  echo?: string;
};

export type GetOtcStatusParams = {
  partnerTxID: string;
  currency: TCurrency;
};

export type GetOtcStatusResponse = {
  txID: string;
  oAuthCode: string;
  paymentMethod: TPaymentMethod;
  /**
   * @deprecated
   */
  status: string;

  txStatus: TTxStatus;
  reason: string;
};

export class MerchantIntegrationOnline {
  _requester: Requester;
  _config: OnlineClientConfig;

  /**
   * Constructor
   */
  constructor(rawConfig: OnlineClientRawConfig, requester: Requester | null = null) {
    this._config = initializeWithConfig(rawConfig);
    this._requester = requester ?? new Requester(this._config);
  }

  /**
   * Init a payment
   */
  chargeInit(payload: ChargeInitParams) {
    const data = {
      ...payload,
      merchantID: this._config.merchantId,
    };

    return this._requester.hmacRequest<ChargeInitResponse, typeof data>({
      data,
      method: 'POST',
      path: this._config.urlSet.ONA_CHARGE_INIT,
    });
  }

  /**
   * Generate Web URL for Authorization
   */
  async generateWebUrl(payload: ChargeInitParams & { state: string; codeVerifier: string }) {
    const { state, codeVerifier, ...chargeInitPayload } = payload;
    const {
      data: { request },
    } = await this.chargeInit(chargeInitPayload);

    const { clientId, redirectUrl, scope, country } = this._config;
    const { currency } = payload;
    const codeChallenge = Utils.base64URLEncode(Utils.hashSha256ToBase64(codeVerifier));

    const params = {
      acr_values: `consent_ctx:countryCode=${country},currency=${currency}`,
      client_id: clientId,
      code_challenge: codeChallenge,
      code_challenge_method: 'S256',
      nonce: Utils.generateRandomString(16),
      redirect_uri: redirectUrl,
      request,
      response_type: 'code',
      scope,
      state,
    };

    return [this._config.envVars.baseUrl, `/grabid/v1/oauth2/authorize?`, new URLSearchParams(params).toString()].join(
      '',
    );
  }

  /**
   * Get OAuth token
   */
  oauth2Token(payload: GetOAuth2TokenParams) {
    const { code, codeVerifier } = payload;
    const { clientId, clientSecret, redirectUrl } = this._config;

    const data = {
      grant_type: 'authorization_code',
      client_id: clientId,
      client_secret: clientSecret,
      code_verifier: codeVerifier,
      redirect_uri: redirectUrl,
      code,
    };

    return this._requester.request<GetOAuth2TokenResponse, typeof data>({
      data,
      headers: {
        'Content-Type': 'application/json',
      },
      method: 'POST',
      path: this._config.urlSet.OAUTH_TOKEN,
    });
  }

  /**
   * Complete payments
   */
  chargeComplete({ accessToken, ...data }: ChargeCompleteParams) {
    return this._requester.request<ChargeCompleteResponse, { partnerTxID: string }>({
      path: this._config.urlSet.ONA_CHARGE_COMPLETE,
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
        Date: Utils.getGMTString(),
        'X-GID-AUX-POP': Utils.generatePOPSignature(accessToken, this._config.clientSecret),
        Authorization: `Bearer ${accessToken}`,
      },
      data,
    });
  }

  /**
   * Get Partner Charge Status
   */
  getChargeStatus(payload: GetChargeStatusParams) {
    const { accessToken, currency, partnerTxID } = payload;
    const path = Utils.replaceUrl(this._config.urlSet.ONA_CHARGE_STATUS, {
      partnerTxID,
    });
    return this._requester.request<GetChargeStatusResponse, never>({
      headers: {
        Accept: 'application/json',
        Authorization: `Bearer ${accessToken}`,
        'Content-Type': 'application/json',
        Date: Utils.getGMTString(),
        'X-GID-AUX-POP': Utils.generatePOPSignature(accessToken, this._config.clientSecret),
      },
      method: 'GET',
      query: {
        currency,
      },
      path,
    });
  }

  /**
   * Generate refunds
   */
  refund({ accessToken, ...payload }: RefundParams) {
    const data = {
      ...payload,
      merchantID: this._config.merchantId,
    };

    return this._requester.request<RefundResponse, typeof data>({
      data,
      headers: {
        Accept: 'application/json',
        Authorization: `Bearer ${accessToken}`,
        'Content-Type': 'application/json',
        Date: Utils.getGMTString(),
        'X-GID-AUX-POP': Utils.generatePOPSignature(accessToken, this._config.clientSecret),
      },
      method: 'POST',
      path: this._config.urlSet.ONA_REFUND,
    });
  }

  /**
   * Get refund status
   */
  getRefundStatus({ accessToken, partnerTxID, currency }: GetRefundStatusParams) {
    const path = Utils.replaceUrl(this._config.urlSet.ONA_REFUND_STATUS, {
      partnerTxID,
    });

    return this._requester.request<GetRefundStatusResponse, never>({
      headers: {
        Accept: 'application/json',
        Authorization: `Bearer ${accessToken}`,
        'Content-Type': 'application/json',
        Date: Utils.getGMTString(),
        'X-GID-AUX-POP': Utils.generatePOPSignature(accessToken, this._config.clientSecret),
      },
      method: 'GET',
      query: {
        currency,
      },
      path,
    });
  }

  /**
   * Get One Time Charge Status
   */
  getOtcStatus({ partnerTxID, currency }: GetOtcStatusParams) {
    const path = Utils.replaceUrl(this._config.urlSet.ONA_ONE_TIME_CHARGE_STATUS, {
      partnerTxID,
    });

    return this._requester.hmacRequest<GetOtcStatusResponse, never>({
      method: 'GET',
      query: {
        currency,
      },
      path,
    });
  }

  /**
   * Get SDK Version
   */
  version() {
    return SDK_VERSION;
  }
}
