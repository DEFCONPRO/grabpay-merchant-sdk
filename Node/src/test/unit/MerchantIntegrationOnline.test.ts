import { initializeWithConfig, OnlineClientRawConfig } from '../../configs/Config';
import { MerchantIntegrationOnline } from '../../MerchantIntegrationOnline';
import Requester from '../../utils/Requester';
import * as Utils from '../../utils/Utils';

const partnerTxID = 'rrSwr3vetp9pRONvRsL0cOK6SzFTKU3d'; // generated using `Utils.generateRandomString(32)`
const partnerGroupTxID = '18csg3DVcCQH9vsA0s2G3zz4CAWE13vi'; // generated using `Utils.generateRandomString(32)`

describe('MerchantIntegrationOnline', () => {
  const config: OnlineClientRawConfig = {
    env: 'STG',
    country: 'VN',
    partnerId: 'partner-id',
    partnerSecret: 'partner-secret',
    merchantId: 'merchant-id',
    clientId: 'client-id',
    clientSecret: 'client-secret',
    redirectUrl: 'http://localhost:8888/result',
  };
  const requester = new Requester(initializeWithConfig(config));
  const axiosRequestSpy = jest
    .spyOn(requester._requester, 'request')
    .mockReturnValue(Promise.resolve({ status: 200, headers: {}, data: {} }));
  let client: MerchantIntegrationOnline;

  beforeEach(() => {
    client = new MerchantIntegrationOnline(config, requester);
    jest.useFakeTimers().setSystemTime(0);
  });

  afterEach(() => {
    jest.clearAllMocks();
    jest.useRealTimers();
  });

  describe('chargeInit', () => {
    it('happy path', async () => {
      await client.chargeInit({
        partnerTxID,
        partnerGroupTxID,
        amount: 100000,
        currency: 'VND',
        description: 'testing otc',
        metaInfo: {
          brandName: 'Store1',
        },
      });
      expect(axiosRequestSpy).toHaveBeenCalledTimes(1);
      expect(axiosRequestSpy.mock.calls).toMatchSnapshot();
    });

    it('hiding postpaid method', async () => {
      await client.chargeInit({
        partnerTxID,
        partnerGroupTxID,
        amount: 100000,
        currency: 'VND',
        description: 'testing otc',
        hidePaymentMethods: ['POSTPAID'],
      });
      expect(axiosRequestSpy).toHaveBeenCalledTimes(1);
      expect(axiosRequestSpy.mock.calls).toMatchSnapshot();
    });
  });

  it('generateWebUrl', async () => {
    const generateRandomStringSpy = jest.spyOn(Utils, 'generateRandomString').mockReturnValue('randomNonce');
    const state = 'someSta'; // generated using `Utils.generateRandomString(7)`
    axiosRequestSpy.mockReturnValue(Promise.resolve({ status: 200, headers: {}, data: { request: 'token' } }));
    const generatedUrl = await client.generateWebUrl({
      partnerTxID,
      partnerGroupTxID,
      amount: 100000,
      currency: 'VND',
      description: 'testing otc',
      metaInfo: {
        brandName: 'Store1',
      },
      codeVerifier: 'abc123',
      state,
    });
    expect(generateRandomStringSpy).toHaveBeenCalledTimes(1);
    expect(axiosRequestSpy).toHaveBeenCalledTimes(1);
    expect(axiosRequestSpy.mock.calls).toMatchSnapshot();
    expect(generatedUrl).toMatchSnapshot();
  });

  it('oauth2Token', async () => {
    await client.oauth2Token({
      code: 'testing',
      codeVerifier: 'abc123',
    });
    expect(axiosRequestSpy).toHaveBeenCalledTimes(1);
    expect(axiosRequestSpy.mock.calls).toMatchSnapshot();
  });

  it('chargeComplete', async () => {
    const payload = { partnerTxID, accessToken: 'testing' };
    await client.chargeComplete(payload);
    expect(axiosRequestSpy).toHaveBeenCalledTimes(1);
    expect(axiosRequestSpy.mock.calls).toMatchSnapshot();
  });

  it('getChargeStatus', async () => {
    await client.getChargeStatus({ partnerTxID, currency: 'VND', accessToken: 'testing' });
    expect(axiosRequestSpy).toHaveBeenCalledTimes(1);
    expect(axiosRequestSpy.mock.calls).toMatchSnapshot();
  });

  it('refund', async () => {
    await client.refund({
      accessToken: 'token',
      partnerTxID,
      partnerGroupTxID,
      amount: 100000,
      currency: 'VND',
      description: 'test',
      originTxID: '34234',
    });
    expect(axiosRequestSpy).toHaveBeenCalledTimes(1);
    expect(axiosRequestSpy.mock.calls).toMatchSnapshot();
  });

  it('getOtcStatus', async () => {
    await client.getOtcStatus({
      currency: 'VND',
      partnerTxID,
    });
    expect(axiosRequestSpy).toHaveBeenCalledTimes(1);
    expect(axiosRequestSpy.mock.calls).toMatchSnapshot();
  });
});
