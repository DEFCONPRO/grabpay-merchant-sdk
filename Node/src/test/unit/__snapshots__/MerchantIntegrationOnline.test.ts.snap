// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MerchantIntegrationOnline chargeComplete 1`] = `
Array [
  Array [
    Object {
      "data": "{\\"partnerTxID\\":\\"rrSwr3vetp9pRONvRsL0cOK6SzFTKU3d\\"}",
      "headers": Object {
        "Accept": "application/json",
        "Authorization": "Bearer testing",
        "Content-Type": "application/json",
        "Date": "Thu, 01 Jan 1970 00:00:00 GMT",
        "X-GID-AUX-POP": "eyJzaWciOiJ2WWdpaEI5UE4zc0VQckxnV1FXUkV3c2JFTENRZmV6bndNODBuRGNhSWlZIiwidGltZV9zaW5jZV9lcG9jaCI6MH0",
        "X-Sdk-Country": "VN",
        "X-Sdk-Language": "nodejs",
        "X-Sdk-Signature": "6EB66646DBF103DC114E34AE6E01C261A217A820357C3B08F3D4E7D4475853C7",
        "X-Sdk-Version": "2.0.0",
      },
      "method": "POST",
      "url": "/mocapay/partner/v2/charge/complete",
    },
  ],
]
`;

exports[`MerchantIntegrationOnline chargeInit happy path 1`] = `
Array [
  Array [
    Object {
      "data": "{\\"partnerTxID\\":\\"rrSwr3vetp9pRONvRsL0cOK6SzFTKU3d\\",\\"partnerGroupTxID\\":\\"18csg3DVcCQH9vsA0s2G3zz4CAWE13vi\\",\\"amount\\":100000,\\"currency\\":\\"VND\\",\\"description\\":\\"testing otc\\",\\"metaInfo\\":{\\"brandName\\":\\"Store1\\"},\\"merchantID\\":\\"merchant-id\\"}",
      "headers": Object {
        "Authorization": "partner-id:PvXQnhoUA0JB0zBc7cd8Ci4hiEH/XBVPc++ZhmOL+PE=",
        "Content-Type": "application/json",
        "Date": "Thu, 01 Jan 1970 00:00:00 GMT",
        "X-Sdk-Country": "VN",
        "X-Sdk-Language": "nodejs",
        "X-Sdk-Signature": "6EB66646DBF103DC114E34AE6E01C261A217A820357C3B08F3D4E7D4475853C7",
        "X-Sdk-Version": "2.0.0",
      },
      "method": "POST",
      "url": "/mocapay/partner/v2/charge/init",
    },
  ],
]
`;

exports[`MerchantIntegrationOnline chargeInit hiding postpaid method 1`] = `
Array [
  Array [
    Object {
      "data": "{\\"partnerTxID\\":\\"rrSwr3vetp9pRONvRsL0cOK6SzFTKU3d\\",\\"partnerGroupTxID\\":\\"18csg3DVcCQH9vsA0s2G3zz4CAWE13vi\\",\\"amount\\":100000,\\"currency\\":\\"VND\\",\\"description\\":\\"testing otc\\",\\"hidePaymentMethods\\":[\\"POSTPAID\\"],\\"merchantID\\":\\"merchant-id\\"}",
      "headers": Object {
        "Authorization": "partner-id:gRYBkFXgq6VZwbY0uD2gBKvAgpxYiaLjC2MKKrG2e5w=",
        "Content-Type": "application/json",
        "Date": "Thu, 01 Jan 1970 00:00:00 GMT",
        "X-Sdk-Country": "VN",
        "X-Sdk-Language": "nodejs",
        "X-Sdk-Signature": "6EB66646DBF103DC114E34AE6E01C261A217A820357C3B08F3D4E7D4475853C7",
        "X-Sdk-Version": "2.0.0",
      },
      "method": "POST",
      "url": "/mocapay/partner/v2/charge/init",
    },
  ],
]
`;

exports[`MerchantIntegrationOnline generateWebUrl 1`] = `
Array [
  Array [
    Object {
      "data": "{\\"partnerTxID\\":\\"rrSwr3vetp9pRONvRsL0cOK6SzFTKU3d\\",\\"partnerGroupTxID\\":\\"18csg3DVcCQH9vsA0s2G3zz4CAWE13vi\\",\\"amount\\":100000,\\"currency\\":\\"VND\\",\\"description\\":\\"testing otc\\",\\"metaInfo\\":{\\"brandName\\":\\"Store1\\"},\\"merchantID\\":\\"merchant-id\\"}",
      "headers": Object {
        "Authorization": "partner-id:PvXQnhoUA0JB0zBc7cd8Ci4hiEH/XBVPc++ZhmOL+PE=",
        "Content-Type": "application/json",
        "Date": "Thu, 01 Jan 1970 00:00:00 GMT",
        "X-Sdk-Country": "VN",
        "X-Sdk-Language": "nodejs",
        "X-Sdk-Signature": "6EB66646DBF103DC114E34AE6E01C261A217A820357C3B08F3D4E7D4475853C7",
        "X-Sdk-Version": "2.0.0",
      },
      "method": "POST",
      "url": "/mocapay/partner/v2/charge/init",
    },
  ],
]
`;

exports[`MerchantIntegrationOnline generateWebUrl 2`] = `"https://stg-paysi.moca.vn/grabid/v1/oauth2/authorize?acr_values=consent_ctx%3AcountryCode%3DVN%2Ccurrency%3DVND&client_id=client-id&code_challenge=bKE9UspwyIPg8LsQHkJaiehiTeUdstI5JZOvaoQRgJA&code_challenge_method=S256&nonce=randomNonce&redirect_uri=http%3A%2F%2Flocalhost%3A8888%2Fresult&request=token&response_type=code&scope=payment.vn.one_time_charge&state=someSta"`;

exports[`MerchantIntegrationOnline getChargeStatus 1`] = `
Array [
  Array [
    Object {
      "headers": Object {
        "Accept": "application/json",
        "Authorization": "Bearer testing",
        "Content-Type": "application/json",
        "Date": "Thu, 01 Jan 1970 00:00:00 GMT",
        "X-GID-AUX-POP": "eyJzaWciOiJ2WWdpaEI5UE4zc0VQckxnV1FXUkV3c2JFTENRZmV6bndNODBuRGNhSWlZIiwidGltZV9zaW5jZV9lcG9jaCI6MH0",
        "X-Sdk-Country": "VN",
        "X-Sdk-Language": "nodejs",
        "X-Sdk-Signature": "6EB66646DBF103DC114E34AE6E01C261A217A820357C3B08F3D4E7D4475853C7",
        "X-Sdk-Version": "2.0.0",
      },
      "method": "GET",
      "params": Object {
        "currency": "VND",
      },
      "url": "/moca/partner/v2/charge/rrSwr3vetp9pRONvRsL0cOK6SzFTKU3d/status",
    },
  ],
]
`;

exports[`MerchantIntegrationOnline getOtcStatus 1`] = `
Array [
  Array [
    Object {
      "headers": Object {
        "Authorization": "partner-id:fbbC6OTohQOiiNGCn3qcJmrA9o0esi7jMdi0Hrf43P0=",
        "Content-Type": "application/json",
        "Date": "Thu, 01 Jan 1970 00:00:00 GMT",
        "X-Sdk-Country": "VN",
        "X-Sdk-Language": "nodejs",
        "X-Sdk-Signature": "6EB66646DBF103DC114E34AE6E01C261A217A820357C3B08F3D4E7D4475853C7",
        "X-Sdk-Version": "2.0.0",
      },
      "method": "GET",
      "params": Object {
        "currency": "VND",
      },
      "url": "/mocapay/partner/v2/one-time-charge/rrSwr3vetp9pRONvRsL0cOK6SzFTKU3d/status",
    },
  ],
]
`;

exports[`MerchantIntegrationOnline oauth2Token 1`] = `
Array [
  Array [
    Object {
      "data": "{\\"grant_type\\":\\"authorization_code\\",\\"client_id\\":\\"client-id\\",\\"client_secret\\":\\"client-secret\\",\\"code_verifier\\":\\"abc123\\",\\"redirect_uri\\":\\"http://localhost:8888/result\\",\\"code\\":\\"testing\\"}",
      "headers": Object {
        "Content-Type": "application/json",
        "X-Sdk-Country": "VN",
        "X-Sdk-Language": "nodejs",
        "X-Sdk-Signature": "6EB66646DBF103DC114E34AE6E01C261A217A820357C3B08F3D4E7D4475853C7",
        "X-Sdk-Version": "2.0.0",
      },
      "method": "POST",
      "url": "/grabid/v1/oauth2/token",
    },
  ],
]
`;

exports[`MerchantIntegrationOnline refund 1`] = `
Array [
  Array [
    Object {
      "data": "{\\"partnerTxID\\":\\"rrSwr3vetp9pRONvRsL0cOK6SzFTKU3d\\",\\"partnerGroupTxID\\":\\"18csg3DVcCQH9vsA0s2G3zz4CAWE13vi\\",\\"amount\\":100000,\\"currency\\":\\"VND\\",\\"description\\":\\"test\\",\\"originTxID\\":\\"34234\\",\\"merchantID\\":\\"merchant-id\\"}",
      "headers": Object {
        "Accept": "application/json",
        "Authorization": "Bearer token",
        "Content-Type": "application/json",
        "Date": "Thu, 01 Jan 1970 00:00:00 GMT",
        "X-GID-AUX-POP": "eyJzaWciOiJIZG5DcHFubDc5bWVWMmtQbWFLNTNHSEkwRFFmWW1DSU5LVGlnc1dzZ0VnIiwidGltZV9zaW5jZV9lcG9jaCI6MH0",
        "X-Sdk-Country": "VN",
        "X-Sdk-Language": "nodejs",
        "X-Sdk-Signature": "6EB66646DBF103DC114E34AE6E01C261A217A820357C3B08F3D4E7D4475853C7",
        "X-Sdk-Version": "2.0.0",
      },
      "method": "POST",
      "url": "/mocapay/partner/v2/refund",
    },
  ],
]
`;
